core/utils/fork-tree/src/lib.rs

```diff
@@ -305,6 +307,79 @@ impl<H, N, V> ForkTree<H, N, V> where
 		}
 	}
 
+	/// Finalize a node in the tree and all its ancestors. The given function
+	/// `is_descendent_of` should return `true` if the second hash (target) is
+	// a descendent of the first hash (base).
+	pub fn finalize_with_ancestors<F, E>(
```

Comment:
I think this is very similar to what we do in `prune` method.

Reply:
Hmm. If I call `prune()`, for example, for block N in [test storage](https://github.com/paritytech/substrate/pull/3301/files#diff-a7d85380647b5b4236cad6aee45caa96R657), it leaves M node. If called for H, the H node is left in the storage. This isn't what expected from `finalize_with_ancestors()`. Probably it could be refactored to support both cases (or we could call `prune()` from within `finalize_with_ancestors()`), but I haven't found a straightforward way to do that.